tests/functional/s3-versioning.sh

```diff
@@ -4,46 +4,322 @@ AWS="aws --endpoint-url http://localhost:5000 --no-verify-ssl"
 ALL_OBJ_VERS=$(${AWS} s3api list-object-versions --bucket "${BUCKET}")
-echo "$ALL_OBJ_VERS" | grep "${OBJ1_1_MD5}"
-echo "$ALL_OBJ_VERS" | grep "${OBJ1_2_MD5}"
+NB_DELETE_MARKERS=$(jq -r ".DeleteMarkers" <<< "$ALL_OBJ_VERS" | jq length)
+NB_VERSIONS=$(jq -r ".Versions" <<< "$ALL_OBJ_VERS" | jq length)
+[[ "$NB_DELETE_MARKERS" -eq "0" ]]
+[[ "$NB_VERSIONS" -eq "2" ]]
+OBJ_1_KEY=$(jq -r ".Versions[1].Key|tostring" <<< "$ALL_OBJ_VERS")
+OBJ_1_ID=$(jq -r ".Versions[1].VersionId|tonumber" <<< "$ALL_OBJ_VERS")
+OBJ_1_MD5=$(jq -r ".Versions[1].ETag|tostring" <<< "$ALL_OBJ_VERS")
+OBJ_1_ISLATEST=$(jq -r ".Versions[1].IsLatest|tostring" <<< "$ALL_OBJ_VERS")
+OBJ_2_KEY=$(jq -r ".Versions[0].Key|tostring" <<< "$ALL_OBJ_VERS")
+OBJ_2_ID=$(jq -r ".Versions[0].VersionId|tonumber" <<< "$ALL_OBJ_VERS")
+OBJ_2_MD5=$(jq -r ".Versions[0].ETag|tostring" <<< "$ALL_OBJ_VERS")
+OBJ_2_ISLATEST=$(jq -r ".Versions[0].IsLatest|tostring" <<< "$ALL_OBJ_VERS")
+[[ "$OBJ_1_KEY" == "obj" ]]
+[[ "$OBJ_2_KEY" == "obj" ]]
+[[ "$OBJ_1_EXPECTED_ID" -eq "$OBJ_1_ID" ]]
+[[ "$OBJ_1_ID" -lt "$OBJ_2_ID" ]]
+[[ "$OBJ_1_MD5" == "\"$OBJ_1_EXPECTED_MD5\"" ]]
+[[ "$OBJ_2_MD5" == "\"$OBJ_2_EXPECTED_MD5\"" ]]
+[[ "$OBJ_1_ISLATEST" == "false" ]]
+[[ "$OBJ_2_ISLATEST" == "true" ]]
+OBJ_2_EXPECTED_ID=$OBJ_2_ID
 
-# The first of the list is the latest version
-OBJ1_1_ID=$(jq ".Versions[1].VersionId|tonumber" <<< "$ALL_OBJ_VERS")
-OBJ1_2_ID=$(jq ".Versions[0].VersionId|tonumber" <<< "$ALL_OBJ_VERS")
+echo "Fetching current version, and checking"
+${AWS} s3 cp "s3://${BUCKET}/obj" obj
+[[ $(md5sum obj | cut -d ' ' -f 1) == "$OBJ_2_EXPECTED_MD5" ]]
 
-echo "Deleting the most recent version"
-${AWS} s3api delete-object --bucket "${BUCKET}" --key "obj1" --version-id "${OBJ1_2_ID}"
+echo "Fetching objects versions, and checking all versions appear"
+if ${AWS} s3api get-object --bucket "${BUCKET}" --key obj --version-id "$OBJ_0_EXPECTED_ID" obj; then
+    false
+fi
+OBJ_1_META=$(${AWS} s3api get-object --bucket "${BUCKET}" --key obj --version-id "$OBJ_1_EXPECTED_ID" obj)
+OBJ_1_MD5=$(jq -r ".ETag|tostring" <<< "$OBJ_1_META")
+[[ "$OBJ_1_MD5" == "\"$OBJ_1_EXPECTED_MD5\"" ]]
+[[ $(md5sum obj | cut -d ' ' -f 1) == "$OBJ_1_EXPECTED_MD5" ]]
+OBJ_2_META=$(${AWS} s3api get-object --bucket "${BUCKET}" --key obj --version-id "$OBJ_2_EXPECTED_ID" obj)
+OBJ_2_MD5=$(jq -r ".ETag|tostring" <<< "$OBJ_2_META")
+[[ "$OBJ_2_MD5" == "\"$OBJ_2_EXPECTED_MD5\"" ]]
+[[ $(md5sum obj | cut -d ' ' -f 1) == "$OBJ_2_EXPECTED_MD5" ]]
+
+echo "*** Putting a delete marker ***"
+${AWS} s3 rm "s3://${BUCKET}/obj"
+
+echo "Listing current version, and checking"
+CUR_VERS=$(${AWS} s3api list-objects --bucket "${BUCKET}")
+[[ -z "$CUR_VERS" ]]
+
+echo "Listing objects versions, and checking all versions appear"
+ALL_OBJ_VERS=$(${AWS} s3api list-object-versions --bucket "${BUCKET}")
+NB_DELETE_MARKERS=$(jq -r ".DeleteMarkers" <<< "$ALL_OBJ_VERS" | jq length)
+NB_VERSIONS=$(jq -r ".Versions" <<< "$ALL_OBJ_VERS" | jq length)
```

Comment:
No need to pipe to another `jq` process.
```suggestion
NB_VERSIONS=$(jq -r ".Versions|length" <<< "$ALL_OBJ_VERS")
```